File: module-services/src/main/java/eu/geekhome/services/automation/DeviceAutomationUnit.java
package eu.geekhome.services.automation;


import eu.geekhome.services.hardware.Connectible;
import eu.geekhome.services.hardware.Port;

import java.util.Calendar;

public abstract class DeviceAutomationUnit<R> extends BlocksTargetAutomationUnit implements IDeviceAutomationUnit<R> {
    private boolean _isConnected = true;

    protected boolean isConnected() {
        return _isConnected;
    }

    public abstract EvaluationResult buildEvaluationResult();

    protected boolean arePortsConnected(Calendar now, Port... ports) {
        for (Port port : ports) {
            if (port instanceof Connectible) {
                Connectible connectible = (Connectible)port;
                boolean isConnected = connectible.checkIfConnected(now.getTimeInMillis());

                if (!isConnected) {
                    return false;
                }
            }
        }

        return true;
    }

    public abstract Port[] getUsedPorts();

    @Override
    public void calculate(Calendar now) throws Exception {
        //if (arePortsConnected(now, getUsedPorts())) {
        //    _isConnected = true;
            calculateInternal(now);
        //} else {
        //    _isConnected = false;
        //}
    }

    protected abstract void calculateInternal(Calendar now) throws Exception;
}

File: module-services/src/main/java/eu/geekhome/services/automation/DeviceAutomationUnit.kt
package eu.geekhome.services.automation

import eu.geekhome.services.R
import java.util.*

abstract class DeviceAutomationUnit<T>(
    var lastEvaluation: EvaluationResult<T>
) {
    abstract val usedPortsIds: Array<String>
    abstract val recalculateOnTimeChange: Boolean
    abstract val recalculateOnPortUpdate: Boolean

    abstract fun calculateInternal(now: Calendar)

    fun calculate(now: Calendar) {
        return try {
            calculateInternal(now)
        } catch (ex: Exception) {
            lastEvaluation = evaluateAsAutomationError(ex)
        }
    }

    private fun evaluateAsAutomationError(ex: java.lang.Exception): EvaluationResult<T> {
        return EvaluationResult(
            interfaceValue = R.error_automation,
            error =  ex)
    }
}